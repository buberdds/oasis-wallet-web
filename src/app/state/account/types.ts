import { Transaction } from 'app/state/transaction/types'
import { ErrorPayload } from 'types/errors'
import { StringifiedBigInt } from 'types/StringifiedBigInt'

export interface BalanceDetails {
  available: StringifiedBigInt | null
  debonding: StringifiedBigInt | null
  delegations: StringifiedBigInt | null
  total: StringifiedBigInt | null
}

export interface Allowance {
  address: string
  amount: StringifiedBigInt
}

export interface Account extends BalanceDetails {
  address: string
  allowances?: Allowance[]
}

/* --- STATE --- */
export interface AccountState extends Account {
  loading: boolean
  accountError?: ErrorPayload
  transactions: Transaction[]
  transactionsError?: ErrorPayload
}
