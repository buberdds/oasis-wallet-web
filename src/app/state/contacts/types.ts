export interface Contact {
  address: string
  name: string
}

/* --- STATE --- */
export interface ContactsState {
  [key: string]: Contact
}
