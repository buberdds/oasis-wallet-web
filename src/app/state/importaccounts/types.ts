import { ErrorPayload } from 'types/errors'
import { WalletBalance, WalletType } from '../wallet/types'

/* --- STATE --- */
export interface ImportAccountsListAccount {
  address: string
  balance: WalletBalance
  path: number[]
  privateKey?: string
  publicKey: string
  selected: boolean
  type: WalletType
}

export enum ImportAccountsStep {
  OpeningUSB = 'opening_usb',
  LoadingAccounts = 'loading_accounts',
  LoadingBalances = 'loading_balances',
  Done = 'done',
}

export interface ImportAccountsState {
  accounts: ImportAccountsListAccount[]
  deviceRequested: boolean
  error?: ErrorPayload
  showAccountsSelectionModal: boolean
  step?: ImportAccountsStep
}
