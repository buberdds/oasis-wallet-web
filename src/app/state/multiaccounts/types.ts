import { ErrorPayload } from 'types/errors'
import { WalletBalance } from '../wallet/types'

/* --- STATE --- */
export interface MultiAccountsListAccount {
  publicKey: string
  address: string
  path: number[]
  balance: WalletBalance
  selected: boolean
}

export enum MultiAccountsStep {
  OpeningUSB = 'opening_usb',
  LoadingAccounts = 'loading_accounts',
  LoadingBalances = 'loading_balances',
  Done = 'done',
}

export interface MultiAccountsState {
  accounts: MultiAccountsListAccount[]
  error?: ErrorPayload
  mnemonic?: string
  step?: MultiAccountsStep
}
