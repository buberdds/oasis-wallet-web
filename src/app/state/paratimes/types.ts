import { StringifiedBigInt } from 'types/StringifiedBigInt'
import { ErrorPayload } from 'types/errors'
import { ParaTime } from '../../../config'

export enum TransactionFormSteps {
  TransferType = 'transferType',
  ParaTimeSelection = 'paraTimeSelection',
  TransactionRecipient = 'transactionRecipient',
  TransactionAmount = 'transactionAmount',
  TransactionConfirmation = 'transactionConfirmation',
  TransactionSummary = 'transactionSummary',
  TransactionError = 'transactionError',
}

export enum TransactionTypes {
  Deposit = 'deposit',
  Withdraw = 'withdraw',
}

export interface TransactionForm {
  amount: string
  confirmation: boolean
  feeAmount: string
  feeGas: string
  paraTime?: ParaTime
  privateKey: string
  recipient: string
  type: TransactionTypes | undefined
}

export interface ParaTimesState {
  balance: StringifiedBigInt
  isLoading: boolean
  transactionError?: ErrorPayload
  transactionForm: TransactionForm
  transactionFormStep: TransactionFormSteps
}

export type Runtime = {
  address: string
  id: string
  decimals: number
}

export type ParaTimeTransaction = Pick<
  TransactionForm,
  'amount' | 'privateKey' | 'feeAmount' | 'feeGas' | 'recipient' | 'type'
>
