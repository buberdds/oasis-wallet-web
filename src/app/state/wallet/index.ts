import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import { LedgerAccount } from '../ledger/types'

import { AddWalletPayload, BalanceUpdatePayload, Wallet, WalletBalance, WalletState } from './types'

export const initialState: WalletState = {
  wallets: {},
  isOpen: false,
  selectedWallet: undefined,
}

const slice = createSlice({
  name: 'wallet',
  initialState,
  reducers: {
    openWalletFromMnemonic(state, action: PayloadAction<string>) {},
    openWalletFromPrivateKey(state, action: PayloadAction<string>) {},
    openWalletsFromLedger(state, action: PayloadAction<LedgerAccount[]>) {},
    selectWallet(state, action: PayloadAction<number>) {},
    closeWallet(state, action: PayloadAction<void>) {},
    fetchWallet(state, action: PayloadAction<Wallet>) {},
    updateBalance(state, action: PayloadAction<BalanceUpdatePayload>) {
      Object.assign(state.wallets[action.payload.walletId].balance, action.payload.balance)
    },
    walletSelected(state, action: PayloadAction<number | undefined>) {
      state.selectedWallet = action.payload
    },
    addWallet(state, action: PayloadAction<AddWalletPayload>) {},
    walletOpened(state, action: PayloadAction<Wallet>) {
      const newWallet = action.payload
      state.wallets[newWallet.id] = Object.assign({}, newWallet)
      state.isOpen = true
    },
    walletClosed(state, action: PayloadAction<void>) {
      // Revert to initial state
      Object.assign(state, initialState)
    },
    walletLoaded(state, action: PayloadAction<WalletBalance>) {
      state.isOpen = true
      // state.wallets[state.selectedWallet!].balance = action.payload
    },
  },
})

export const { actions: walletActions } = slice

export default slice.reducer
