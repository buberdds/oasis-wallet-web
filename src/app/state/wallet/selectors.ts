import { createSelector } from '@reduxjs/toolkit'

import { RootState } from 'types'
import { initialState } from '.'

const selectSlice = (state: RootState) => state.wallet || initialState
export const selectActiveWallet = createSelector([selectSlice], state => {
  if (state.selectedWallet == null) {
    return undefined
  } else {
    return state.wallets[state.selectedWallet]
  }
})

export const selectWallets = createSelector([selectSlice], state => state.wallets ?? {})
export const selectWalletsPublicKeys = createSelector([selectWallets], wallets =>
  Object.values(wallets).map(w => w.publicKey),
)
export const selectAddress = createSelector([selectActiveWallet], wallet => wallet?.address)
export const selectPublicKey = createSelector([selectActiveWallet], wallet => wallet?.publicKey ?? '')
export const selectBalance = createSelector([selectActiveWallet], wallet => wallet?.balance)
export const selectType = createSelector([selectActiveWallet], wallet => wallet?.type)
export const selectIsOpen = createSelector([selectSlice], wallet => wallet.isOpen)
