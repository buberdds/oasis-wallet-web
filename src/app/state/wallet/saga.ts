import { hdkey } from '@oasisprotocol/client'
import { PayloadAction } from '@reduxjs/toolkit'
import { hex2uint, parseRpcBalance, publicKeyToAddress, shortPublicKey, uint2hex } from 'app/lib/helpers'
import nacl from 'tweetnacl'
import { call, fork, put, select, take, takeEvery, takeLatest } from 'typed-redux-saga'

import { walletActions, initialState } from '.'
import { LedgerAccount } from '../ledger/types'
import { getOasisNic } from '../network/saga'
import { transactionActions } from '../transaction'
import { selectAddress, selectWallets } from './selectors'
import { AddWalletPayload, Wallet, WalletType } from './types'

// Ensure a unique walletId per opened wallet
// Maybe we should switch to something like uuid later
let walletId = 0

/**
 * Opened wallet saga
 * Will later be used to sign arbitrary messages
 */
export function* walletSaga() {}

export function* rootWalletSaga() {
  // Wait for an openWallet action (Mnemonic, Private Key, Ledger) and add them if requested
  yield* takeEvery(walletActions.openWalletFromPrivateKey, openWalletFromPrivateKey)
  yield* takeEvery(walletActions.openWalletFromMnemonic, openWalletFromMnemonic)
  yield* takeEvery(walletActions.openWalletsFromLedger, openWalletsFromLedger)
  yield* takeEvery(walletActions.addWallet, addWallet)

  // Reload balance of matching wallets when a transaction occurs
  yield* fork(reloadBalanceOnTransaction)
  yield* takeEvery(walletActions.fetchWallet, loadWallet)

  // Allow switching between wallets
  yield* takeLatest(walletActions.selectWallet, selectWallet)

  // Start the wallet saga in parallel
  yield* fork(walletSaga)

  // Listen to closeWallet
  yield* takeEvery(walletActions.closeWallet, closeWallet)
}

export function* getBalance(publicKey: Uint8Array) {
  const nic = yield* call(getOasisNic)
  const short = yield* call(shortPublicKey, publicKey)
  const account = yield* call([nic, nic.stakingAccount], {
    height: 0,
    owner: short,
  })

  return parseRpcBalance(account)
}

function* getWalletByAddress(address: string) {
  const wallets = yield* select(selectWallets)
  const wallet = Object.values(wallets).find(w => w.address === address)

  return wallet ? wallet : undefined
}
/**
 * Take multiple ledger accounts that we want to open
 */
export function* openWalletsFromLedger({ payload: accounts }: PayloadAction<LedgerAccount[]>) {
  for (const [index, account] of accounts.entries()) {
    yield* put(
      walletActions.addWallet({
        id: walletId++,
        address: account.address,
        publicKey: account.publicKey,
        type: WalletType.Ledger,
        balance: account.balance,
        path: account.path,
        selectImmediately: index === 0,
      }),
    )
  }
}

export function* openWalletFromPrivateKey({ payload: privateKey }: PayloadAction<string>) {
  const type = WalletType.PrivateKey
  const publicKeyBytes = nacl.sign.keyPair.fromSecretKey(hex2uint(privateKey)).publicKey
  const walletAddress = yield* call(publicKeyToAddress, publicKeyBytes)
  const publicKey = uint2hex(publicKeyBytes)
  const balance = yield* call(getBalance, publicKeyBytes)

  yield* put(
    walletActions.addWallet({
      id: walletId++,
      address: walletAddress,
      publicKey,
      privateKey,
      type: type!,
      balance,
      selectImmediately: true,
    }),
  )
}

export function* openWalletFromMnemonic({ payload: mnemonic }: PayloadAction<string>) {
  const signer = yield* call(hdkey.HDKey.getAccountSigner, mnemonic)
  const privateKey = uint2hex(signer.secretKey)
  const type = WalletType.Mnemonic
  const publicKeyBytes = signer.publicKey
  const publicKey = uint2hex(publicKeyBytes)

  const walletAddress = yield* call(publicKeyToAddress, publicKeyBytes!)
  const balance = yield* call(getBalance, publicKeyBytes)

  yield* put(
    walletActions.addWallet({
      id: walletId++,
      address: walletAddress,
      publicKey,
      privateKey,
      type: type!,
      balance,
      selectImmediately: true,
    }),
  )
}

/**
 * Adds a wallet to the existing wallets
 * If the wallet exists already, do nothingg
 * If it has "selectImmediately", we select it immediately
 */
export function* addWallet({ payload: newWallet }: PayloadAction<AddWalletPayload>) {
  yield* put(walletActions.walletSelected(initialState.selectedWallet))
  const existingWallet = yield* call(getWalletByAddress, newWallet.address)
  if (!existingWallet) {
    yield* put(walletActions.walletOpened(newWallet))
  }

  const walletId = existingWallet ? existingWallet.id : newWallet.id

  if (newWallet.selectImmediately) {
    yield* put(walletActions.selectWallet(walletId))
  }
}

export function* closeWallet() {
  yield* put(walletActions.walletClosed())
}

export function* selectWallet({ payload: index }: PayloadAction<number>) {
  yield* put(walletActions.walletSelected(index))
}

function* loadWallet(action: PayloadAction<Wallet>) {
  const wallet = action.payload
  const balance = yield* call(getBalance, hex2uint(wallet.publicKey))
  yield* put(
    walletActions.updateBalance({
      walletId: wallet.id,
      balance,
    }),
  )
}

/**
 * When a transaction is done, and it is related to the account we currently have in state
 * refresh the data.
 */
function* reloadBalanceOnTransaction() {
  while (true) {
    const { payload } = yield* take(transactionActions.transactionSent)
    if (payload.type !== 'transfer') {
      // @TODO: This should be done for other types of transactions too
      return
    }

    const from = yield* select(selectAddress)
    const to = payload.to

    const wallets = yield* select(selectWallets)
    const matchingWallets = Object.values(wallets).filter(w => w.address === to || w.address === from)
    for (const wallet of matchingWallets) {
      yield* put(walletActions.fetchWallet(wallet))
    }
  }
}
