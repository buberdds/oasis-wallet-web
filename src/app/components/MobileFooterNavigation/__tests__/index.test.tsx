import * as React from 'react'
import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'
import { ThemeProvider } from 'styles/theme/ThemeProvider'
import { Wallet } from 'app/state/wallet/types'

import { MobileFooterNavigation, MobileFooterNavigationProps } from '..'

const renderComponent = (store: any, { isAccountOpen, isMobile }: MobileFooterNavigationProps) =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <ThemeProvider>
          <MobileFooterNavigation isAccountOpen={isAccountOpen} isMobile={isMobile} />
        </ThemeProvider>
      </MemoryRouter>
    </Provider>,
  )

describe('<isAccountOpen />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore({
      wallet: {
        isOpen: true,
        selectedWallet: 0,
        wallets: {
          0: {
            address: 'dummy',
            id: 1,
          } as Wallet,
        },
      },
    })
  })

  it('should render component for mobile and when account is open', () => {
    renderComponent(store, { isAccountOpen: true, isMobile: true })

    expect(screen.queryByTestId('mobile-footer-navigation')).toMatchSnapshot()
  })

  it('should not render component for non mobile', () => {
    renderComponent(store, { isAccountOpen: true, isMobile: false })

    expect(screen.queryByTestId('mobile-footer-navigation')).not.toBeInTheDocument()
  })

  it('should not render component when account is not open', () => {
    renderComponent(store, { isAccountOpen: false, isMobile: true })

    expect(screen.queryByTestId('mobile-footer-navigation')).not.toBeInTheDocument()
  })
})
