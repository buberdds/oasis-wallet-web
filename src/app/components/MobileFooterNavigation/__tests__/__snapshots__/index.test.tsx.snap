// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<isAccountOpen /> should render component for mobile and when account is open 1`] = `
.c4 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c4 g {
  fill: inherit;
  stroke: inherit;
}

.c4 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c4 *[stroke*="#"],
.c4 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c4 *[fill-rule],
.c4 *[FILL-RULE],
.c4 *[fill*="#"],
.c4 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin: 6px;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  font-size: 14px;
  line-height: 20px;
}

.c1 {
  background-color: #FFFFFF;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  height: 4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  box-shadow: 0px 0px 1px #EDEDED;
  border-top: solid 1px #11111108;
  font-size: 0.875rem;
  line-height: 1;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

@media only screen and (max-width:768px) {
  .c3 {
    margin: 3px;
  }
}

<div
  class="c0 c1"
  data-testid="mobile-footer-navigation"
>
  <a
    href="/account/dummy"
  >
    <span
      class="c2"
    >
      <span
        class="c3"
      >
        <svg
          aria-label="Money"
          class="c4"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 16c0-1.105-3.134-2-7-2s-7 .895-7 2 3.134 2 7 2 7-.895 7-2zM2 16v4.937C2 22.077 5.134 23 9 23s7-.924 7-2.063V16M9 5c-4.418 0-8 .895-8 2s3.582 2 8 2M1 7v5c0 1.013 3.582 2 8 2M23 4c0-1.105-3.1-2-6.923-2-3.824 0-6.923.895-6.923 2s3.1 2 6.923 2S23 5.105 23 4zm-7 12c3.824 0 7-.987 7-2V4M9.154 4v10.166M9 9c0 1.013 3.253 2 7.077 2C19.9 11 23 10.013 23 9"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </span>
      <span
        class="c5"
      >
        menu.wallet
      </span>
    </span>
  </a>
  <a
    href="/account/dummy/stake"
  >
    <span
      class="c2"
    >
      <span
        class="c3"
      >
        <svg
          aria-label="LineChart"
          class="c4"
          viewBox="0 0 24 24"
        >
          <path
            d="m1 16 7-7 5 5L23 4M0 22h23.999M16 4h7v7"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </span>
      <span
        class="c5"
      >
        menu.stake
      </span>
    </span>
  </a>
</div>
`;
